### PR TITLE
unstably allow constants to refer to statics and read from immutable statics

### PR DESCRIPTION
**Pull request by RalfJung**
I am not aware of any fundamental reason why we cannot allow constants to mention statics. What we really need is that constants do not *read from* statics that can change their value:
- This would break the principle that "constants behave as-if their expression was inlined everywhere and executed at runtime". This is enforced by halting const-eval interpretation when a read from a mutable global occurs.
- When building a valtree we want to be sure that the constant and everything it refers to is truly immutable. This is enforced by aborting valtree construction when a read from a mutable global occurs.

r? @oli-obk -- if you are okay with experimenting with this feature, I will create a tracking issue.
Based on and blocked on https://github.com/rust-lang/rust/pull/119044; only the last commit is new.

### CODE REVIEW

**Review by oli-obk — APPROVED** [review]
fine by me, as long as the tracking issue mentions this new source of monomorphization time errors as an open question to put to T-lang

File: tests/ui/feature-gates/feature-gate-const-refs-to-static.stderr [hunk 1]
```diff
@@ -0,0 +1,28 @@
+error[E0013]: constants cannot refer to statics
+  --> $DIR/feature-gate-const-refs-to-static.rs:4:19
+   |
+LL | const C1: &i32 = &S;
+   |                   ^
+   |
+   = help: consider extracting the value of the `static` to a `const`, and referring to that
```
Comment (fmease) [inline tests/ui/feature-gates/feature-gate-const-refs-to-static.stderr:7]: You've probably already planned on adding that but Imma comment anyway:

```diff
+    = note: see issue #NNNNNN <https://github.com/rust-lang/rust/issues/NNNNNN> for more information
```

And for `tcx.sess.is_nightly_build()`:

```diff
+ help: add `#![feature(const_refs_to_static)]` to the crate attributes to enable
```

Comment (RalfJung) [reply tests/ui/feature-gates/feature-gate-const-refs-to-static.stderr:7]: That seems like an issue with our const stability checks -- not related to this PR.

File: compiler/rustc_const_eval/src/transform/check_consts/ops.rs [hunk 2]
```diff
@@ -580,16 +580,21 @@ impl<'tcx> NonConstOp<'tcx> for StaticAccess {
...
             Status::Allowed
         } else {
-            Status::Forbidden
+            Status::Unstable(sym::const_refs_to_static)
         }
     }
 
     fn build_error(&self, ccx: &ConstCx<'_, 'tcx>, span: Span) -> DiagnosticBuilder<'tcx> {
-        ccx.dcx().create_err(errors::StaticAccessErr {
+        let mut err = feature_err(
```
Comment (fmease) [inline compiler/rustc_const_eval/src/transform/check_consts/ops.rs:588]: :heart:

File: compiler/rustc_const_eval/src/transform/check_consts/ops.rs [hunk 3]
```diff
@@ -580,16 +580,21 @@ impl<'tcx> NonConstOp<'tcx> for StaticAccess {
...
+            sym::const_refs_to_static,
             span,
-            kind: ccx.const_kind(),
-            teach: ccx.tcx.sess.teach(&error_code!(E0013)).then_some(()),
-        })
+            format!("referencing statics in {}s is unstable", ccx.const_kind(),),
+        );
+        err
+            .note("`static` and `const` variables can refer to other `const` variables. A `const` variable, however, cannot refer to a `static` variable.")
+            .help("To fix this, the value can be extracted to a `const` and then used.");
```
Comment (fmease) [inline compiler/rustc_const_eval/src/transform/check_consts/ops.rs:596]:
```suggestion
            .note("`static` and `const` variables can refer to other `const` variables; a `const` variable, however, cannot refer to a `static` variable")
            .help("to fix this, the value can be extracted to a `const` and then used");
```

https://rustc-dev-guide.rust-lang.org/diagnostics.html?highlight=lowercase#diagnostic-output-style-guide (subitem 2)

File: tests/ui/feature-gates/feature-gate-const-refs-to-static.stderr (hunk 1 above)
Comment (RalfJung) [reply tests/ui/feature-gates/feature-gate-const-refs-to-static.stderr]: Ah I have to put the feature gate into `build_error` as well.

**Comment by oli-obk** [comment]
seems fishy to allow valtrees to read from immutable statics. That means we meaningfully lose the static's identity during rountdrip.

**Comment by RalfJung** [comment]
Valtrees are about the value. How would the static's identity matter?

**Comment by oli-obk** [comment]
I think it is perfectly valid for users to assume that a static's address will not get erased during intermediate stages. Even if it is not valid to assume that a generic const ever constains any meaningful addresses.

```rust
static FOO: usize = 42;
const BAR: &usize = &FOO;
fn foo<const X: &usize>() {
    if X.addr() == FOO.addr() {
        println!("activating special mode");
    }
}

foo::<BAR>();
```

**Comment by RalfJung** [comment]
That seems extremely sketchy to me, const generics do not have identity.

I don't think this is any worse than
```rust
const BAR: (&usize, &usize) = {
  let val = &0; (val, val)
};
fn foo<const X: (&usize, &usize)>() {
    if X.0.addr() == X.1.addr() {
        println!("activating special mode");
    }
}

foo::<BAR>();
```

**Comment by oli-obk** [comment]
> That seems extremely sketchy to me, const generics do not have reference identity.

yea, but losing the identity of static items anywhere seems also sketchy to me 😆

Unless we say that immutable frozen static items are equivalent to const items, except that they are guaranteed to have a memory location (and only one of them at that).

**Comment by RalfJung** [comment]
So, my plan here was to file an issue about the existing ways in which valtree roundtrips break reference equality, and make this just yet another item for that issue -- a concern that certainly needs to block stabilizing either of these features, but doesn't have to block experimenting with a nightly feature I think.

However, I wasn't able to reproduce the issues that I expect a roundtrip would cause; probably this means my code doesn't actually cause a rondtrip. I'd appreciate help understanding what is going on here. See this [Zulip thread](https://rust-lang.zulipchat.com/#narrow/stream/260443-project-const-generics/topic/Forcing.20a.20valtree.20roundtrip).

**Comment by Darksonn** [comment]
Even if you're not able to trigger it, can we still move forward? Could we file the issue without a reproducer, then merge this for experimentation in nightly?

**Comment by RalfJung** [comment]
@rust-lang/project-const-generics This PR interacts in subtle ways with values of reference types that are round-tripped through a const generic argument -- it allows them to contain references to immutable statics, whose contents will be copied and a fresh reference be created as part of passing the const generic argument (which is how all reference-typed const generics work).

Personally I feel like since this is the interaction of two unstable features, we can land this for experimentation, and figure out how to resolve this issue as we get closer to stabilizing either reference-typed const generics or constants allowed to refer to statics. Please let us know if you have any objections to that plan.

**Comment by BoxyUwU** [comment]
Given that references in const generic types already do not retain any knowledge about addresses I think this ought to be fine.

Does this PR handle `static FOO: UnsafeCell<u32>` correctly, we dont want the `u32` value to be able to feed into a const generic value. Do we forbid accessing immutable statics with types that have interior mutability?

**Comment by RalfJung** [comment]
> Does this PR handle static FOO: UnsafeCell<u32> correctly, we dont want the u32 value to be able to feed into a const generic value. Do we forbid accessing immutable statics with types that have interior mutability?

Yes, "immutable" for this PR means "`static` without interior mutability".

File: tests/ui/consts/miri_unleashed/const_refers_to_static.rs [hunk 4]
```diff
@@ -17,17 +17,15 @@ const READ_INTERIOR_MUT: usize = {
 static mut MUTABLE: u32 = 0;
 const READ_MUT: u32 = unsafe { MUTABLE }; //~ERROR evaluation of constant value failed
 
-const REF_INTERIOR_MUT: &usize = { //~ ERROR undefined behavior to use this value
-//~| encountered a reference pointing to a static variable
+// Not actually reading from anything mutable, so these are fine.
```
Comment (RalfJung) [inline tests/ui/consts/miri_unleashed/const_refers_to_static.rs:20]: I need to fix this comment.

**Comment by matthiaskrgr** [comment]
@bors r-
needs rebase / blessing
https://github.com/rust-lang/rust/pull/120709#issuecomment-1930492768

**Pull request merged**